router/usecase/pools/routable_cw_alloy_transmuter_pool.go

```diff
@@ -0,0 +1,166 @@
+// Transmuter pool allows no slippage swaps. For v3, the ratio of token in to token out is dependent on the normalization factor.
+// Returns error if:
+// - the underlying chain pool set on the routable pool is not of transmuter type
+// - the token in amount is greater than the balance of the token in
+// - the token in amount is greater than the balance of the token out
+//
+// Note that balance validation does not apply to alloyed asset since it can be minted or burned by the pool.
+func (r *routableAlloyTransmuterPoolImpl) CalculateTokenOutByTokenIn(ctx context.Context, tokenIn sdk.Coin) (sdk.Coin, error) {
+	tokenOutAmt, err := r.calcTokenOutAmt(tokenIn, r.TokenOutDenom)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	tokenOutAmtInt := osmomath.NewIntFromBigInt(tokenOutAmt.TruncateInt().BigInt())
+
+	// Validate token out balance if not alloyed
+	if r.TokenOutDenom != r.AlloyTransmuterData.AlloyedDenom {
+		if err := validateTransmuterBalance(tokenOutAmtInt, r.Balances, r.TokenOutDenom); err != nil {
+			return sdk.Coin{}, err
+		}
+	}
+
+	return sdk.Coin{Denom: r.TokenOutDenom, Amount: tokenOutAmtInt}, nil
+}
+
+// GetTokenOutDenom implements RoutablePool.
+func (r *routableAlloyTransmuterPoolImpl) GetTokenOutDenom() string {
+	return r.TokenOutDenom
+}
+
+// String implements sqsdomain.RoutablePool.
+func (r *routableAlloyTransmuterPoolImpl) String() string {
+	return fmt.Sprintf("pool (%d), pool type (%d) Transmuter with alloyed denom, pool denoms (%v), token out (%s)", r.ChainPool.PoolId, poolmanagertypes.CosmWasm, r.GetPoolDenoms(), r.TokenOutDenom)
+}
+
+// ChargeTakerFeeExactIn implements sqsdomain.RoutablePool.
+// Returns tokenInAmount and does not charge any fee for transmuter pools.
+func (r *routableAlloyTransmuterPoolImpl) ChargeTakerFeeExactIn(tokenIn sdk.Coin) (inAmountAfterFee sdk.Coin) {
+	return tokenIn
+}
+
+// GetTakerFee implements sqsdomain.RoutablePool.
+func (r *routableAlloyTransmuterPoolImpl) GetTakerFee() math.LegacyDec {
+	return r.TakerFee
+}
+
+// SetTokenOutDenom implements sqsdomain.RoutablePool.
+func (r *routableAlloyTransmuterPoolImpl) SetTokenOutDenom(tokenOutDenom string) {
+	r.TokenOutDenom = tokenOutDenom
+}
+
+// CalcSpotPrice implements sqsdomain.RoutablePool.
+func (r *routableAlloyTransmuterPoolImpl) CalcSpotPrice(ctx context.Context, baseDenom string, quoteDenom string) (osmomath.BigDec, error) {
+	return r.calcTokenOutAmt(sdk.Coin{Denom: baseDenom, Amount: osmomath.OneInt()}, quoteDenom)
+}
+
+// IsGeneralizedCosmWasmPool implements sqsdomain.RoutablePool.
+func (*routableAlloyTransmuterPoolImpl) IsGeneralizedCosmWasmPool() bool {
+	return false
+}
+
+// GetCodeID implements sqsdomain.RoutablePool.
+func (r *routableAlloyTransmuterPoolImpl) GetCodeID() uint64 {
+	return r.ChainPool.CodeId
+}
+
+func (r *routableAlloyTransmuterPoolImpl) findNormalizationFactors(tokenInDenom, tokenOutDenom string) (osmomath.Int, osmomath.Int, error) {
```

Comment:
Would be great to unit test each smaller method individually - the happy paths and major error/edge cases 🙏